### PR TITLE
Reload configuration when file changes

### PR DESCRIPTION
**Pull request by icetrain**
Another version of "reload config on change" that checks mtime of the configuration file and tries to load the new configuration, it that fails it keeps running on the previous settings.

Solves #29

### CODE REVIEW

File: collector/src/main/java/io/prometheus/jmx/JmxCollector.java [hunk 1]
```diff
@@ -39,83 +40,97 @@
...
+    private static class Config {
+      String jmxUrl = "";
+      String username = "";
+      String password = "";
+      boolean lowercaseOutputName;
+      boolean lowercaseOutputLabelNames;
+      List<ObjectName> whitelistObjectNames = new ArrayList<ObjectName>();
+      List<ObjectName> blacklistObjectNames = new ArrayList<ObjectName>();
+      ArrayList<Rule> rules = new ArrayList<Rule>();
+      long lastUpdate = System.currentTimeMillis();
```
Comment (brian-brazil) [inline collector/src/main/java/io/prometheus/jmx/JmxCollector.java]: This doesn't feel right. I think you should set this to the file mtime.

File: collector/src/main/java/io/prometheus/jmx/JmxCollector.java [hunk 2]
```diff
@@ -39,83 +40,97 @@
...
+        config = loadConfig((Map<String, Object>)new Yaml().load(yamlConfig));
+    }
+
+    private void reloadConfig() {
+      try {
+        FileReader fr = new FileReader(configFile);
+        Map<String, Object> newYamlConfig = (Map<String, Object>)new Yaml().load(fr);
+        config = loadConfig(newYamlConfig);
+      } catch (Exception e) {
+        LOGGER.severe("Configuration reload failed: " + e.toString());
```
Comment (brian-brazil) [inline collector/src/main/java/io/prometheus/jmx/JmxCollector.java]: Maybe a increment a metric here? You can use direction instrumentation.

File: collector/src/main/java/io/prometheus/jmx/JmxCollector.java (hunk 1 above)
Comment (brian-brazil) [inline collector/src/main/java/io/prometheus/jmx/JmxCollector.java]: A metric with the file mtime would be a good idea too.

File: collector/src/main/java/io/prometheus/jmx/JmxCollector.java [hunk 3]
```diff
@@ -39,83 +40,97 @@
...
+      }
     }
-    private JmxCollector(Map<String, Object> config) throws MalformedObjectNameException {
-        if(config == null) {  //Yaml config empty, set config to empty map.
-            config = new HashMap<String, Object>();
+
+    private Config loadConfig(Map<String, Object> yamlConfig) throws MalformedObjectNameException {
+        Config cfg = new Config();
+
+        if(yamlConfig == null) {  //Yaml config empty, set config to empty map.
```
Comment (brian-brazil) [inline collector/src/main/java/io/prometheus/jmx/JmxCollector.java]: Space after //

File: collector/src/main/java/io/prometheus/jmx/JmxCollector.java [hunk 4]
```diff
@@ -39,83 +40,97 @@
...
     }
+
     public JmxCollector(String yamlConfig) throws MalformedObjectNameException {
-        this((Map<String, Object>)new Yaml().load(yamlConfig));
+        config = loadConfig((Map<String, Object>)new Yaml().load(yamlConfig));
+    }
+
+    private void reloadConfig() {
+      try {
+        FileReader fr = new FileReader(configFile);
```
Comment (iggyzap) [inline collector/src/main/java/io/prometheus/jmx/JmxCollector.java]: That's a resource leak.

**Comment by icetrain** [comment]
Update with suggestions, let me know if you would like a rebase.

**Comment by brian-brazil** [comment]
Thanks both.

**Pull request merged**